### PR TITLE
feat(linter): implement @typescript-eslint/prefer-as-const

### PR DESCRIPTION
**Pull request by Thiry1**
implement @typescript-eslint/prefer-as-const
Related issue: #503.

### CODE REVIEW

File: crates/oxc_linter/Cargo.toml [hunk 1]
```diff
@@ -28,7 +28,7 @@ regex       = { workspace = true }
 rustc-hash  = { workspace = true }
 phf         = { workspace = true, features = ["macros"] }
 num-traits  = { workspace = true }
-itertools  = { workspace = true }
+itertools   = { workspace = true }
```
Comment (Thiry1) [inline crates/oxc_linter/Cargo.toml:31]: I ran `just fmt` on root.

File: crates/oxc_linter/src/rules/typescript/prefer_as_const.rs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        "class foo { bar; }",
+        "class foo { bar: string = 'baz'; }",
+        "class foo { bar: number = 1; }",
+        "class foo { bar = 'baz' as const; }",
+        "class foo { bar = 2 as const; }",
+        "class foo { get bar(): 'bar' {} set bar(bar: 'bar') {} }",
+        "class foo { bar = () => 'bar' as const; }",
+        "type BazFunction = () => 'baz'; class foo { bar: BazFunction = () => 'bar'; }",
+        "class foo { bar(): void {} }",
+        // NOTE: OXC does not parse these format yet.
```
Comment (Thiry1) [inline crates/oxc_linter/src/rules/typescript/prefer_as_const.rs:165]: Is my understanding correct?

**Comment by Thiry1** [comment]
I am a beginner in Rust, so the code quality may be low.

File: crates/oxc_linter/src/rules/typescript/prefer_as_const.rs (hunk 2 above)
Comment (Boshen) [reply crates/oxc_linter/src/rules/typescript/prefer_as_const.rs:159]: https://www.typescriptlang.org/docs/handbook/jsx.html#the-as-operator

We are parsing everything as tsx, you can keep them commented out.

**Review by Boshen — DISMISSED** [review]

File: crates/oxc_linter/src/rules/typescript/prefer_as_const.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    ts_type: &TSType,
+    initial_value_expression: &Expression,
+    ctx: &LintContext,
+    can_fix: bool,
+) {
+    if let TSType::TSLiteralType(literal_type) = &ts_type {
+        let error_span = match &literal_type.literal {
+            TSLiteral::StringLiteral(string_literal) => match initial_value_expression {
+                Expression::StringLiteral(initial_string) => {
+                    if string_literal.value.to_string().eq(&initial_string.value.to_string()) {
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/typescript/prefer_as_const.rs:88]: You don't need `to_string` here, it creates a memory allocation.

**Comment by Boshen** [comment]
I ran `just fmt` on main as well, you can remove them from this PR.

**Comment by Thiry1** [comment]
@Boshen Thank you for review.I finished fixing it

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Awesome work! Thank you!

**Pull request merged**